### PR TITLE
Migrate project metadata from setup.py to pyproject.toml following PEP621

### LINKED ISSUE #1836: Migrate settings from setup.py/setup.cfg to pyproject.toml
As mentioned in https://github.com/GenericMappingTools/pygmt/issues/1797#issuecomment-1067056606, it's better to remove `setup.py`/`setup.cfg` and move all project metadata and packaging/buiding to `pyproject.toml`.

Here are some related resources I found on the internet:

### PEPs

- [PEP 517 – A build-system independent format for source trees](https://peps.python.org/pep-0517/)
- [PEP 518 – Specifying Minimum Build System Requirements for Python Projects](https://peps.python.org/pep-0518/)
- [PEP 621 – Storing project metadata in pyproject.toml](https://peps.python.org/pep-0621/)
- [PEP 631 – Dependency specification in pyproject.toml based on PEP 508](https://peps.python.org/pep-0631/)

#### General

- https://chadsmith.dev/python-packaging/
- https://bernat.tech/posts/pep-517-and-python-packaging/
- https://bernat.tech/posts/pep-517-518/

#### Setuptools

- https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html
- https://github.com/pypa/setuptools/issues/1688
- https://discuss.python.org/t/help-testing-experimental-features-in-setuptools/13821
- https://github.com/pypa/setuptools/issues/2816
- 
### Poetry

- https://python-poetry.org/
- https://github.com/mtkennerly/poetry-dynamic-versioning
- https://github.com/GenericMappingTools/pygmt/pull/1055
- https://github.com/python-poetry/roadmap/issues/3

### Flit

- https://flit.pypa.io/en/latest/
- https://github.com/pypa/flit/issues/257

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Fixes #1836.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pyproject.toml
Comment (weiji14) [inline pyproject.toml:12]: Probably should stop using @leouieda's email here. Don't we have a proper PyGMT dev email now @meghanrjones?

File: setup.py
Comment (weiji14) [inline setup.py:21]: Is it possible to put this `PACKAGE_DATA` line in `pyproject.toml`?

File: setup.py
Comment (weiji14) [inline setup.py:3]: Maybe these lines need to be kept to make the `black` linter happy.

File: setup.py
Comment (weiji14) [inline setup.py:3]: I haven't looked at PEP621 (or the other PEPs) too closely, but could you explain why a `setup.py` file is still needed?

**Review by weiji14 — COMMENTED** [review]
Thanks @seisman! Just some quick comments for now. There seems to be 166 in CI due to missing baseline images for some reason, is there something missing?

File: setup.py
Comment (seisman) [reply setup.py]: Done in 7cff5908.

File: setup.py
Comment (seisman) [reply setup.py:10]: The main reason is setuptools with a `pyproject.toml` configuration file still doesn't support editable installs (i.e., `pip install -e`). 

See https://github.com/pypa/setuptools/issues/2816 and the note at the top of  this page (https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).

File: pyproject.toml
Comment (seisman) [inline pyproject.toml:2]: Please note that changes in this PR require setuptools>=61. 

[setuptools v61.0.0](https://setuptools.pypa.io/en/latest/history.html#v61-0-0) was released on Mar. 24, 2020. This version starts to [support storing project metadata in `pyproject.toml` following PEP621](https://github.com/pypa/setuptools/issues/1688) but it's still experimental and may change in future versions. I think we should not merge this PR unless the `pyproject.toml` support is stable.

File: pyproject.toml
Comment (meghanrjones) [reply pyproject.toml:12]: Not that I am aware of.

Two options:
1. Setup a listserve with the maintainers' email addresses. We could probably do this through UH but there are some sustainability issues there.
2. Register a new gmail (or other email provided) account and share login information.

I might actually prefer option 1 if we can find a stable alternative to UH listserv, because I would rather have emails forwarded to an existing address than have another email address to check.

File: pyproject.toml
Comment (weiji14) [reply pyproject.toml:2]: Yes, let's wait and see first. I've had to pin to setuptools=58.4.0 on one of my projects 3 weeks ago because one of the dependencies didn't like setuptools v60.9. There might be some other packages pinning to a lower setuptools version too.

File: setup.py
Comment (imba-tjd) [reply setup.py:10]: How about `pip install -e`?

File: pyproject.toml
Comment (leouieda) [reply pyproject.toml:12]: > Probably should stop using @leouieda's email here.

+1 to that 🙂 

@meghanrjones I recently made one for Fatiando for this. You can setup auto forwarding in Gmail to other addresses so you wouldn't have to check it.

**Comment by leouieda** [comment]
Having recently gone through all of this, there are 2 options that would work well here:

1. Use `setup.cfg` and `pyproject.toml` (see [Pooch](https://github.com/fatiando/pooch)) so you don't need `setup.py` and have everything in text format already (though in 2 files). Once setuptools support is stable, move the config to `pyproject.toml` only. This can take a year or more since setuptools underpins so many projects. So keeping this PR open until then will make it stale.
2. Migrate from setuptools to [flit](https://flit.pypa.io/en/latest/) which already supports `pyproject.toml`. It's a different built backend and is only compatible with pure Python packages. Since PyGMT doesn't have compiled code, this would work. But it would also make it impossible to distribute libgmt with PyPI wheels in the future (which would be awesome but difficult to achieve). 

Either way, getting rid of `setup.py` means that source and wheel packages have to be built by a different tool. I'm using [build](https://pypa-build.readthedocs.io/en/latest/) which is pretty straight forward (replace `python setup.py sdist bdist_wheel` with `python -m build . `).

File: pyproject.toml
Comment (meghanrjones) [reply pyproject.toml:12]: Thanks for the info. I created pygmt.team@gmail.com.

File: pyproject.toml
Comment (seisman) [inline pyproject.toml:11]:
```suggestion
authors = [{name = "The PyGMT Developers", email = "pygmt.team@gmail.com"}]
```

File: pyproject.toml
Comment (seisman) [inline pyproject.toml:12]:
```suggestion
maintainers = [{name = "The PyGMT Developers", email = "pygmt.team@gmail.com"}]
```

Comment (seisman) [reply pyproject.toml]: I've updated the email address in cc9d8e0d23de4f888a7c17466572d21639b4fd02.

**Comment by seisman** [comment]
> 1. Use `setup.cfg` and `pyproject.toml` (see [Pooch](https://github.com/fatiando/pooch)) so you don't need `setup.py` and have everything in text format already (though in 2 files). Once setuptools support is stable, move the config to `pyproject.toml` only. This can take a year or more since setuptools underpins so many projects. So keeping this PR open until then will make it stale.

We are also trying to get rid of the `setup.cfg` file in https://github.com/GenericMappingTools/pygmt/pull/1847. So moving from `setup.py` to `setup.cfg` is not an option.

> 2. Migrate from setuptools to [flit](https://flit.pypa.io/en/latest/) which already supports `pyproject.toml`. It's a different built backend and is only compatible with pure Python packages. Since PyGMT doesn't have compiled code, this would work. But it would also make it impossible to distribute libgmt with PyPI wheels in the future (which would be awesome but difficult to achieve).

I also noticed `flit`, but it seems flit doesn't support a dynamic version like what `setuptools_scm` does (https://github.com/pypa/flit/issues/257).

> Either way, getting rid of `setup.py` means that source and wheel packages have to be built by a different tool. I'm using [build](https://pypa-build.readthedocs.io/en/latest/) which is pretty straight forward (replace `python setup.py sdist bdist_wheel` with `python -m build . `).

Yes, we have switched to `build` (https://github.com/GenericMappingTools/pygmt/pull/1823) a few weeks ago. 🎆

**Comment by leouieda** [comment]
Ah right, if keeping setup.cfg is out of the question then it would a matter of waiting until setuptools support is stable. It underpins so many packages that generating conflicts with it will likely be a nightmare.

File: pyproject.toml
Comment (seisman) [reply pyproject.toml:45]: Perhaps we can set the version string to `0.0.0` instead.

Comment (seisman) [reply pyproject.toml:45]: PR in https://github.com/GenericMappingTools/pygmt/pull/1945.

File: setup.py
Comment (seisman) [inline setup.py:1]: setup.py is no longer required since setuptools 64.0.0 (https://setuptools.pypa.io/en/latest/userguide/development_mode.html)

Comment (weiji14) [reply setup.py]: Cool, it seems like setuptools v64.0.0 enabled non setup.py editable installs!

On the developer side, this seems ok to remove setup.py, but are there any potential side effects for users not having such a new setuptools version? Or to put it another way, do we want to remove setup.py for PyGMT v0.8.0, or wait another minor version (e.g. v0.9.0)?

Comment (seisman) [reply setup.py]: > are there any potential side effects for users not having such a new setuptools version?

I have setuptools 62.6.0 installed. Checking out this branch and running `make install` and `make package` work well for me.

When running `make package`, I see the following messages:
```
$ make package 
python -m build --sdist --wheel
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=64, setuptools_scm[toml]>=6.2)
* Getting dependencies for sdist...
/tmp/build-env-ld441dbo/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py:108: _BetaConfiguration: Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
  warnings.warn(msg, _BetaConfiguration)
running egg_info
writing pygmt.egg-info/PKG-INFO
writing dependency_links to pygmt.egg-info/dependency_links.txt
writing requirements to pygmt.egg-info/requires.txt
writing top-level names to pygmt.egg-info/top_level.txt
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE.txt'
adding license file 'AUTHORS.md'
adding license file 'AUTHORSHIP.md'
writing manifest file 'pygmt.egg-info/SOURCES.txt'
* Building sdist...
/tmp/build-env-ld441dbo/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py:108: _BetaConfiguration: Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
  warnings.warn(msg, _BetaConfiguration)
running sdist
running egg_info
writing pygmt.egg-info/PKG-INFO
writing dependency_links to pygmt.egg-info/dependency_links.txt
writing requirements to pygmt.egg-info/requires.txt
writing top-level names to pygmt.egg-info/top_level.txt
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE.txt'
adding license file 'AUTHORS.md'
adding license file 'AUTHORSHIP.md'
writing manifest file 'pygmt.egg-info/SOURCES.txt'
running check
...
```
It seems build dependencies like setuptools>=64 are automatically installed in isolated environments and do not affect users' installations.

File: doc/maintenance.md
Comment (seisman) [inline doc/maintenance.md:170]: We have the following command at line 280:
```
python setup.py --long-description | rst2html.py --no-raw > index.html
```
After removing `setup.py`, the command is no longer valid.

File: setup.py
Comment (seisman) [reply setup.py:21]: Done in f257a361b2ef870b0f9d159835fd5cbac065c5cb.

File: doc/maintenance.md
Comment (seisman) [reply doc/maintenance.md:170]: I've replaced the command with:
```
rst2html.py --no-raw README.rst > index.html
```
in 7553f0b4.

**Comment by seisman** [comment]
> There seems to be 166 in CI due to missing baseline images for some reason, is there something missing?

That's because `make package` doesn't include the baseline images.

File: pyproject.toml
Comment (weiji14) [inline pyproject.toml:26]: Small suggestion to have this as multi-line to have a cleaner diff when we bump the minimum NumPy version.

```suggestion
dependencies = [
    "numpy>=1.20",
    "pandas",
    "xarray",
    "netCDF4",
    "packaging"
]
```

Comment (weiji14) [inline pyproject.toml:26]: Thinking of whether to declare optional dependencies (https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#dependencies-optional-dependencies). Maybe do in a separate PR.

Motivation would be to enable something like `pip install pygmt[all]` that includes dependencies like geopandas.

**Review by weiji14 — APPROVED** [review]
Great work, keeping PyGMT nice and modern :smiley: Just some very minor suggestions, otherwise looks good to me.

File: setup.py
Comment (maxrjones) [reply setup.py]: A downside here is that `make install` no longer works without internet access, even if setuptools>=64 is installed in the environment. I'll do some more searching for a workaround but if anyone knows a solution that would be helpful.

Comment (weiji14) [reply setup.py]: I tried disconnecting and used `pip install --no-deps --no-build-isolation --editable=.` which seems to work offline (presumably it manages to pick up a pre-downloaded `setuptools>=64`). However, it results in a editable version named `pygmt-0.0.0`.

Comment (seisman) [reply setup.py]: > A downside here is that `make install` no longer works without internet access, even if setuptools>=64 is installed in the environment.

It's unrelated to this PR. Offline `make install` worked for pygmt<=0.2.0 and started to fail since v0.3.0, in which we switched from versioneer to setuptools_scm (https://www.pygmt.org/dev/changes.html#id47). So it's likely a problem for setuptools_scm.

> I tried disconnecting and used `pip install --no-deps --no-build-isolation --editable=.` which seems to work offline (presumably it manages to pick up a pre-downloaded `setuptools>=64`). However, it results in a editable version named `pygmt-0.0.0`.

This command gives the correct version for pygmt <= 0.6.0 but results in `pygmt-0.0.0` for pygmt v0.7.0. Not sure if it's caused by changes in https://github.com/GenericMappingTools/pygmt/pull/1945.

File: pyproject.toml
Comment (seisman) [reply pyproject.toml]: Opened a PR https://github.com/GenericMappingTools/pygmt/pull/2069 for it.

File: setup.py
Comment (seisman) [reply setup.py]: @maxrjones Perhaps open an issue report for the offline `pip` issue?

Comment (seisman) [reply setup.py]: I've opened a PR in https://github.com/GenericMappingTools/pygmt/issues/2097.

**Pull request merged**
